docs/root/configuration/http_filters/jwt_authn_filter.rst

```diff
@@ -0,0 +1,214 @@
+.. _config_http_filters_jwt_authn:
+
+JWT Authentication
+==================
+
+This HTTP filter can be used to verify JSON Web Token (JWT). It will verify its signature, expiration time, audiences and issuer. If the JWT verification fails, its request will be rejected. If the JWT verification success, its payload could be forwarded to the upstream for further authorization if desired.
+
+JWKS is needed to verify JWT signatures. They can be specified in the filter config or can be fetched remotely from a JWKS server.
+
+Configuration
+-------------
+
+This HTTP :ref:`filer config <envoy_api_msg_config.filter.http.jwt_authn.v2alpha.JwtAuthentication>` has two fields, *providers* and *rules*. Field *providers* specifies how a JWT should be verified, such as where to extract the token, where to fetch the public key (JWKS) and where to output its payload. Field *rules* specifies matching rules and their requirements. It a request matches a rule, its requirement applies. The requirement specifies which JWT providers should be used.
+
+JwtProvider
+~~~~~~~~~~~
+
+:ref:`JwtProvider <envoy_api_msg_config.filter.http.jwt_authn.v2alpha.JwtProvider>` specifies how a JWT should be verified. It has following fields:
+
+* *issuer*: the principal that issued the JWT, usually a URL or an email address.
+* *audiences*: a list of JWT audiences allowed to access. A JWT containing any of these audiences will be accepted.
+  If not specified, will not check audiences in the JWT.
+* *local_jwks*: fetch JWKS in local data source, either in a local file or embedded in the inline string.
+* *remote_jwks*: fetch JWKS from a remote HTTP server, also specify cache duration.
+* *forward*: if true, JWT will be forwarded to the upstream.
+* *from_headers*: extract JWT from HTTP headers.
+* *from_params*: extract JWT from query parameters.
+* *forward_payload_header*: forward the JWT payload in the specified HTTP header.
+
+Default Extract Location
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+If *from_headers* and *from_params* is empty,  the default location to extract JWT is from HTTP header::
+
+  Authorization: Bearer <token>
+
+If fails to extract a JWT from above header, then check query parameter key *access_token* as this example::
+
+  /path?access_token=<JWT>
+
+In the :ref:`filter config <envoy_api_msg_config.filter.http.jwt_authn.v2alpha.JwtAuthentication>`, *providers* is a map, to map *provider_name* to a :ref:`JwtProvider <envoy_api_msg_config.filter.http.jwt_authn.v2alpha.JwtProvider>`. The *provider_name* has to be unique, it is referred in the `JwtRequirement <envoy_api_msg_config.filter.http.jwt_authn.v2alpha.JwtRequirement>` in its *provider_name* field.
```

Comment:
The \*provider_name\* **must** be unique?

Reply:
Done